janus_server/src/trace.rs

```diff
@@ -1,40 +1,88 @@
-use tracing_subscriber::{fmt, layer::SubscriberExt, EnvFilter, Registry};
+//! Configures a tracing subscriber for Janus.
 
+use atty::{self, Stream};
+use serde::{Deserialize, Serialize};
+use tracing_log::LogTracer;
+use tracing_subscriber::{fmt::Layer, layer::SubscriberExt, EnvFilter, Registry};
+
+/// Errors from initializing trace subscriber.
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     #[error("tracing error: {0}")]
-    HttpClient(#[from] tracing::subscriber::SetGlobalDefaultError),
+    SetGlobalTracingSubscriber(#[from] tracing::subscriber::SetGlobalDefaultError),
+    #[error("logging error: {0}")]
+    SetGlobalLogger(#[from] tracing_log::log_tracer::SetLoggerError),
+}
+
+/// Configuration for the tracing subscriber.
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TraceConfiguration {
+    /// If true, uses a [`tracing_subscriber::fmt::TestWriter`] to capture trace
+    /// events when running tests
+    #[serde(default)]
+    pub use_test_writer: bool,
+    /// If true OR if stdout is a tty, trace events are output in JSON format by
+    /// [`tracing_subscriber::fmt::format::Json`]. Otherwise, trace events are
+    /// output in pretty format by [`tracing_subscriber::fmt::format::Pretty`].
+    #[serde(default)]
+    pub force_json_output: bool,
 }
 
-/// Configures and installs a tracing subscriber
-pub fn install_subscriber() -> Result<(), Error> {
-    // Configure a tracing subscriber. The crate emits events using `info!`,
-    // `err!`, etc. macros from crate `tracing`.
-    let fmt_layer = fmt::layer()
+/// Create a base tracing layer with configuration used in all subscribers
+fn base_layer<S>() -> Layer<S> {
+    tracing_subscriber::fmt::layer()
         .with_thread_ids(true)
-        // TODO(#16): take an argument for pretty vs. full vs. compact vs. JSON
-        // output
-        .pretty()
         .with_level(true)
-        .with_target(true);
+        .with_target(true)
+        .with_file(true)
+        .with_line_number(true)
+}
+
+/// Configures and installs a tracing subscriber, to capture events logged with
+/// [`tracing::info`] and the like. Captured events are written to stdout, with
+/// formatting affected by the provided [`TraceConfiguration`].
+pub fn install_subscriber(config: &TraceConfiguration) -> Result<(), Error> {
+    // If stdout is not a tty or if forced by config, output logs as JSON
+    // structures
+    let output_json = atty::isnt(Stream::Stdout) || config.force_json_output;
+
+    let (pretty_layer, json_layer, test_layer) = match (output_json, config.use_test_writer) {
+        (true, false) => (None, Some(base_layer().json()), None),
+        (false, false) => (Some(base_layer().pretty()), None, None),
+        (_, true) => (None, None, Some(base_layer().pretty().with_test_writer())),
+    };
```

Comment:
It's not especially elegant, but I was able to get this working without repeating too much code.